oauth2_http/java/com/google/auth/oauth2/AwsCredentials.java

```diff
@@ -0,0 +1,286 @@
+    region = region.substring(0, region.length() - 1);
+
+    AwsSecurityCredentials credentials = getAwsSecurityCredentials();
+
+    // Generate the signed request to the AWS STS GetCallerIdentity API.
+    Map<String, String> headers = new HashMap<>();
+    headers.put("x-goog-cloud-target-resource", audience);
+
+    AwsRequestSigner signer =
+        AwsRequestSigner.newBuilder(
+                credentials,
+                "POST",
+                ((AwsCredentialSource) credentialSource)
+                    .regionalCredVerificationUrl.replace("{region}", region),
+                region)
+            .setAdditionalHeaders(headers)
+            .build();
+
+    AwsRequestSignature awsRequestSignature = signer.sign();
+    return buildSubjectToken(awsRequestSignature);
+  }
+
+  /** Clones the AwsCredentials with the specified scopes. */
+  @Override
+  public GoogleCredentials createScoped(Collection<String> newScopes) {
+    return new AwsCredentials(
+        transportFactory,
+        audience,
+        subjectTokenType,
+        tokenUrl,
+        tokenInfoUrl,
+        (AwsCredentialSource) credentialSource,
+        serviceAccountImpersonationUrl,
+        quotaProjectId,
+        clientId,
+        clientSecret,
+        newScopes);
+  }
+
+  private String retrieveResource(String url, String resourceName) throws IOException {
+    try {
+      HttpRequestFactory requestFactory = transportFactory.create().createRequestFactory();
+      HttpRequest request = requestFactory.buildGetRequest(new GenericUrl(url));
+      HttpResponse response = request.execute();
+      return response.parseAsString();
+    } catch (IOException e) {
+      throw new IOException(String.format("Failed to retrieve AWS %s.", resourceName), e);
+    }
+  }
+
+  private String buildSubjectToken(AwsRequestSignature signature) {
+    GenericJson headers = new GenericJson();
+    headers.setFactory(OAuth2Utils.JSON_FACTORY);
+
+    Map<String, String> canonicalHeaders = signature.getCanonicalHeaders();
+    for (String headerName : canonicalHeaders.keySet()) {
+      headers.put(headerName, canonicalHeaders.get(headerName));
+    }
+
+    headers.put("Authorization", signature.getAuthorizationHeader());
+    headers.put("x-goog-cloud-target-resource", audience);
+
+    GenericJson token = new GenericJson();
+    token.setFactory(OAuth2Utils.JSON_FACTORY);
+
+    token.put("headers", headers);
+    token.put("method", signature.getHttpMethod());
+    token.put(
+        "url",
+        ((AwsCredentialSource) credentialSource)
+            .regionalCredVerificationUrl.replace("{region}", signature.getRegion()));
+    return token.toString();
+  }
+
+  @VisibleForTesting
+  AwsSecurityCredentials getAwsSecurityCredentials() throws IOException {
+    // Check environment variables for credentials first.
+    String accessKeyId = getEnv("AWS_ACCESS_KEY_ID");
+    String secretAccessKey = getEnv("AWS_SECRET_ACCESS_KEY");
+    String token = getEnv("Token");
```

Comment:
I'm not a fan of how deep in the code this is reading environment variables. Ideally, this is done at credential detection/creation time and not in the flow of fetching the access tokens.

Reply:
Since it's optional here--i.e. the code runs if these variables aren't set--this feels OK to me.

Reply:
I don't mind moving this to the constructor, but AFAICT I wouldn't be able to test that code path. What do you guys think?